src/topology/homotopy/basic.lean

```diff
@@ -0,0 +1,299 @@
+by { ext, simp }
+
+/--
+Given `homotopy_with f₀ f₁ P` and `homotopy_with f₁ f₂ P`, we can define a `homotopy_with f₀ f₂ P`
+by putting the first homotopy on `[0, 1/2]` and the second on `[1/2, 1]`.
+-/
+def trans {f₀ f₁ f₂ : C(X, Y)} (F : homotopy_with f₀ f₁ P) (G : homotopy_with f₁ f₂ P) :
+  homotopy_with f₀ f₂ P :=
+{ to_fun := λ x, if (x.1 : ℝ) ≤ 1/2 then F.extend (2 * x.1) x.2 else G.extend (2 * x.1 - 1) x.2,
+  continuous_to_fun := begin
+    refine continuous_if_le _ _ (continuous.continuous_on _) (continuous.continuous_on _) _,
+    swap 5,
+    { rintros x hx,
+      norm_num [hx] },
+    exact continuous_induced_dom.comp continuous_fst,
+    exact continuous_const,
+    -- TODO: Work out why `continuity` doesn't work here.
+    exact F.continuous.comp
+      ((continuous_proj_Icc.comp (continuous_const.mul
+        (continuous_induced_dom.comp continuous_fst))).prod_mk continuous_snd),
+    exact G.continuous.comp
+      ((continuous_proj_Icc.comp
+      ((continuous_const.mul
+        (continuous_induced_dom.comp continuous_fst)).sub continuous_const)).prod_mk
+          continuous_snd),
+  end,
+  to_fun_zero := λ x, by norm_num,
+  to_fun_one := λ x, by norm_num,
+  prop' := λ t, begin
+    simp only,
+    split_ifs,
+    { exact F.extend_prop _ },
+    { exact G.extend_prop _ },
+  end }
+
+lemma trans_apply {f₀ f₁ f₂ : C(X, Y)} (F : homotopy_with f₀ f₁ P) (G : homotopy_with f₁ f₂ P)
+  (x : I × X) : (F.trans G) x =
+  if h : (x.1 : ℝ) ≤ 1/2 then
+    F (⟨2 * x.1, (unit_interval.mul_pos_mem_iff zero_lt_two).2 ⟨x.1.2.1, h⟩⟩, x.2)
+  else
+    G (⟨2 * x.1 - 1, unit_interval.two_mul_sub_one_mem_iff.2 ⟨(not_le.1 h).le, x.1.2.2⟩⟩, x.2) :=
+show ite _ _ _ = _,
+by split_ifs; { rw [extend, continuous_map.coe_Icc_extend, set.Icc_extend_of_mem], refl }
+
+lemma symm_trans {f₀ f₁ f₂ : C(X, Y)} (F : homotopy_with f₀ f₁ P) (G : homotopy_with f₁ f₂ P) :
+  (F.trans G).symm = G.symm.trans F.symm :=
+begin
+  ext x,
+  simp only [symm_apply, trans_apply],
+  split_ifs with h₁ h₂,
+  { change (x.1 : ℝ) ≤ _ at h₂,
+    change (1 : ℝ) - x.1 ≤ _ at h₁,
+    have ht : (x.1 : ℝ) = 1/2,
+    { linarith },
+    norm_num [ht] },
+  { congr' 2,
+    apply subtype.ext,
+    simp only [unit_interval.coe_symm_eq, subtype.coe_mk],
+    linarith },
+  { congr' 2,
+    apply subtype.ext,
+    simp only [unit_interval.coe_symm_eq, subtype.coe_mk],
+    linarith },
+  { change ¬ (x.1 : ℝ) ≤ _ at h,
+    change ¬ (1 : ℝ) - x.1 ≤ _ at h₁,
+    exfalso, linarith }
+end
+
+end homotopy_with
+
+/--
+A `homotopy f₀ f₁` is defined to be a `homotopy_with` where the predicate is always `true`.
+-/
+abbreviation homotopy (f₀ f₁ : C(X, Y)) := homotopy_with f₀ f₁ (λ f, true)
+
+namespace homotopy
+
+/--
+Given a map `f : C(X, Y)`, we can define a `homotopy f f` by setting `F (t, x) = f x` for all `t`.
+This is defined using `homotopy_with.refl`, but with the proof filled in.
+-/
+def refl (f : C(X, Y)) : homotopy f f := homotopy_with.refl f trivial
```

Comment:
`@[simps]`?

Don't we also need `symm` and `trans` here?

Reply:
Oh right. For `symm` and `trans`, the definitions for `homotopy_with` work as is, so I think I should be able to just add them as `alias`es.